plutus-contract/src/Plutus/Contract/StateMachine.hs

```diff
@@ -517,12 +529,14 @@ mkStep client@StateMachineClient{scInstance} input = do
                         red = Ledger.Redeemer (PlutusTx.toBuiltinData (Scripts.validatorHash typedValidator, Burn))
                         unmint = if isFinal then mustMintValueWithRedeemer red (inv $ SM.threadTokenValueOrZero scInstance) else mempty
                         outputConstraints =
-                            [ OutputConstraint
-                                { ocDatum = stateData newState
-                                  -- Add the thread token value back to the output
-                                , ocValue = stateValue newState <> SM.threadTokenValueOrZero scInstance
-                                }
-                            | not isFinal ]
+                            let newStateValue = stateValue newState <> SM.threadTokenValueOrZero scInstance
+                                missingLovelace = max 0 (Ledger.minAdaTxOut - Ada.fromValue (stateValue newState))
+                             in [ OutputConstraint
+                                    { ocDatum = stateData newState
+                                      -- Add the thread token value back to the output
+                                    , ocValue = newStateValue <> Ada.toValue missingLovelace
+                                  }
+                                | not isFinal && not (Value.isZero newStateValue) ]
```

Comment:
Ah, this is what I was missing. I don't think we want to throw away the outputconstraint when the value is zero, since we still want to propagate the ocDatum.

Reply:
Then what do we do if the output value is zero? Simply adjusting it to 2 Ada doesn't make much sense, since that would mean that I could never use all funds in a script address.

Reply:
And similarly it would not be possible to use all the funds but 1 Ada in a script address right? Maybe we need to discuss this with people who have thought about this more...

Reply:
So currently, we need to adapt contracts like `GameStateMachine` such that if the locked funds are 0, then we need to transition of a final state. Of course, if want all the funds - 1, then the adjusting function will change that to `funds - 2`.